ethereum/contracts/BeefyClient.sol

```diff
@@ -0,0 +1,474 @@
+
+        ValidatorSet memory vset = getValidatorSet(data.validatorSetID);
+
+        // verify that block wait period has passed
+        require(block.number >= data.blockNumber + BLOCK_WAIT_PERIOD, "Block wait period not over");
+
+        return
+            Bitfield.randomNBitsWithPriorCheck(
+                getSeed(data),
+                data.validatorClaimsBitfield,
+                requiredNumberOfSignatures(vset),
+                vset.length
+            );
+    }
+
+    function createInitialBitfield(uint256[] calldata bitsToSet, uint256 length)
+        public
+        pure
+        returns (uint256[] memory)
+    {
+        return Bitfield.createBitfield(bitsToSet, length);
+    }
+
+    /**
+     * @notice Performs the final step in the validation, and then applies the commitment and optional authority update
+     * @param id an identifying value generated in the previous transaction
+     * @param commitment contains the full commitment that was used for the commitmentHash
+     * @param proof a struct containing the data needed to verify all validator signatures
+     */
+    function completeSignatureCommitment(
+        uint256 id,
+        Commitment calldata commitment,
+        ValidatorProof calldata proof
+    ) public {
+        ValidationData storage data = validationData[id];
+
+        bytes32 commitmentHash = verifyCommitment(data, commitment, proof);
+
+        latestMMRRoot = commitment.payload.mmrRootHash;
+        latestBeefyBlock = commitment.blockNumber;
+        emit NewMMRRoot(commitment.payload.mmrRootHash, commitment.blockNumber);
+
+        // Check if commitment signals an authority handover (new validator session)
+        if (commitment.validatorSetId == currentValidatorSet.id + 1) {
+            currentValidatorSet = nextValidatorSet;
+            emit NewSession(
+                commitment.blockNumber,
+                nextValidatorSet.id,
+                nextValidatorSet.root,
+                nextValidatorSet.length
+            );
+        }
+
+        // Obtain a gas refund
+        delete validationData[id];
+
+        emit CommitmentVerified(id, Phase.Final, commitmentHash, msg.sender);
+    }
+
+    function updateValidatorSet(MMRLeaf calldata leaf, MMRProof calldata proof) public {
+        require(leaf.nextAuthoritySetId == nextValidatorSet.id + 1, "Leaf is invalid");
+
+        // Verify that the leaf suppied by the relayer is part of the MMR
+        bytes32 leafHash = keccak256(encodeMMRLeaf(leaf));
+        require(
+            MMRProofVerification.verifyLeafProof(latestMMRRoot, leafHash, proof),
+            "Invalid leaf proof"
+        );
+
+        nextValidatorSet.id = leaf.nextAuthoritySetId;
+        nextValidatorSet.root = leaf.nextAuthoritySetRoot;
+        nextValidatorSet.length = leaf.nextAuthoritySetLen;
+    }
+
+    /* Private Functions */
+
+    function getValidatorSet(uint256 id) internal view returns (ValidatorSet memory) {
```

Comment:
same suggestion as above - can we just use a boolean true/false for next vs current here?

Reply:
obsolete